xla/service/while_loop_analysis_test.cc

```diff
@@ -130,6 +130,71 @@ TEST_F(WhileLoopAnalysisTest, SingleIterationUpperBound) {
   EXPECT_EQ(*ComputeWhileLoopTripCountUpperBound(while_op), 1);
 }
 
+TEST_F(WhileLoopAnalysisTest, SimpleLoopWithCustomCallNonTuple){
+  std::string hlo_string = R"(
+  HloModule SimpleLoop
+  SimpleLoop.body {
+    loop_var.1 = (s32[]{:T(128)}, s32[3]{0}) parameter(0)
+    custom-call.1 = (s32[]{:T(128)}, s32[3]{0}) custom-call(loop_var.1), custom_call_target="CustomCallStart"
```

Comment:
Here, are we assuming that the custom calls at the beginning and end do not modify the loop trip count (especially decrement it)? This seems a bit odd. What is the rationale behind this?

Reply:
Correct they will not modify the tripcount. 
We are guaranteed that also, as these custom-calls are not modifying the while loop trip count logic. If for some reason, the trip count was modified - it would refuse to unroll as static trip count fails. 

You can read the rationale for this in the PR. Some while loop bodies we still want to unroll that have custom-calls at start and end of the while loop body.

Reply:
So in while loop analysis it calls ComputeWhileLoopTripCount (function may be named differently but similar) this computes the trip count of the while loop and if it is static then it allows an unroll.

Reply:
I think the assumption here is that the custom call does not use the passed in loop induction variable in-place. We cannot know what a custom call does, but as long as it does not have in-place parameters, it cannot modify the passed in parameter. To make a custom call have in-place parameters, it needs to be explicitly handled in buffer assignment (e.g. here: https://github.com/openxla/xla/blob/0fc891390264fb85ac822f45c4106c48e1a10ffc/xla/service/gpu/buffer_sharing.cc#L232).

Reply:
The problem here is that we are using the output of this custom call to update it and then we send that value as the updated induction variable. It doesn't matter if the update is in-place or not, because we are using the output of the custom call for the increment operation. For this to be safe, we need to somehow assert that the custom call returns the loop induction variable unchanged. Am I missing something here?

Reply:
Ah, good point, I had not noticed that. I agree, this seems problematic. In fact, if the custom call returns the induction variable unchanged, why make it an output of the custom call in the first place?

Reply:
So I remember now - it is assumed the custom-call will not alter the trip count. This is a special case for Neuron hardware in that we sometimes have custom-calls at the start and end of our while loop body, and we want to still unroll them.

Reply:
A safer solution to this problem in my opinion would to be to add an unsafe pass `replace-while-custom-call-ind-var` right before unrolling (default to off). This pass would replace the uses of the returned induction variable by the original induction variable. This pass would only run for the special case of neuron hardware and will not make the while loop analysis unsafe. For this testcase, the pass would change the HLO to the following:
```
SimpleLoop.body {
    loop_var.1 = (s32[]{:T(128)}, s32[3]{0}) parameter(0)
    custom-call.1 = (s32[]{:T(128)}, s32[3]{0}) custom-call(loop_var.1), custom_call_target="CustomCallStart"
    get-tuple-element.1 = s32[]{:T(128)} get-tuple-element(loop_var.1), index=0
    constant.1 = s32[]{:T(128)} constant(1)
    idx = s32[]{:T(128)} add(get-tuple-element.1, constant.1)
    get-tuple-element.2 = s32[3]{0} get-tuple-element(custom-call.1), index=1
    output = s32[3]{0} add(get-tuple-element.2, get-tuple-element.2)
    custom-call.2 = (s32[]{:T(128)}, s32[3]{0}) custom-call(idx, output), custom_call_target="CustomCallEnd"
    final_output_data = s32[3]{0} get-tuple-element(custom-call.2), index=1
    ROOT tuple = (s32[]{:T(128)}, s32[3]{0})tuple(idx, final_output_data)
  }
```
With such a pass, while loop unrolling should still unroll the loop. Will this work for you @ptoulme-aws ? What do you think about this solution @akuegel @fhoushmand ?

Reply:
For my use case, I needed the custom-calls to be the first and last instructions of the while loop body, so this would not work.

I wanted to be able to just run the pass that inserts these custom-calls (proprietary pass) immediately before unroll but that PR was not accepted. Perhaps you could add a boolean flag - allow custom-call induction var?

Reply:
We should not make the while analysis unsafe (even behind a flag) because many passes rely on this pass and the trip count calculated. If the proprietary passes require the custom calls to be the first and last instructions to be custom-calls, then maybe there should be passes for both these tasks:

1. Before unroll pass, one pass should replace the uses of output of custom call to the original loop induction variable (as suggested above)
2. After unroll pass, another pass should start reusing the output of custom call as the loop induction variable and make the custom-calls the first and last instructions of each iteration of the unrolled loop so that the proprietary passes can run on them.

Both these passes would be internal (or default to off in upstream), and so this would not make while loop analysis or unrolling inaccurate upstream.

Reply:
How is a flag unsafe if the caller guarantees the custom-calls are just pass through for the trip count var? 
Those solutions are all cumbersome. A flag is a simple solution.

Reply:
You can default flag to off obviously. My point is - it is not unsafe if the person using the flags know why they are using it. 
Default to False.

Reply:
Thanks for catching this issue, I'm ashamed that I didn't notice it!

> How is a flag unsafe if the caller guarantees the custom-calls are just pass through for the trip count var? Those solutions are all cumbersome. A flag is a simple solution.
>
> How is a flag unsafe if the caller guarantees the custom-calls are just pass through for the trip count var?
Those solutions are all cumbersome. A flag is a simple solution.

We do not generally wish to support flags in XLA, unless there's a particularly good reason to do that—and this does not seem like one. Flags are cumbersome, and significantly increase our testing & maintenance surface area.

The solution you suggest @shraiysh sounds reasonable. If there is no way to make this safe in the general case, and we insist that Neuron needs to specialize this use case, then the dangerous transformation needs to be gated for Neuron hardware only.

Failing that, we'll probably need to revert this, since the code as it is right now is unsafe.

Reply:
Do we have the steps forward? Do we have a design for a safe implementation of this in while loop analysis?

Reply:
Looking at your proposal 

1. Before unroll pass, one pass should replace the uses of output of custom call to the original loop induction variable (as suggested above)
2. After unroll pass, another pass should start reusing the output of custom call as the loop induction variable and make the custom-calls the first and last instructions of each iteration of the unrolled loop so that the proprietary passes can run on them.

Reply:
For 1. This assumes custom-call does not change the loop induction var. This is effectively same assumption as the current pass.

Reply:
2. This also assumes it does not change loop induction var.

Reply:
Your proposal is effectively "if loop induction var is output of custom-call, change graph to use loop induction var of custom-call for unroll purposes to force a guranteed static trip count, but then put the loop induction var back in the custom-call."

Am I understanding correctly? This does not seem safe and is not transparent to users

Reply:
The core transform that neuron hardware requires here (according to the description above) is unsafe. In the proposed solution, the unsafe part of the transform is decoupled from the while loop analysis and does not affect the compilation for other hardware and users. These two unsafe passes would be internal to neuron hardware. 

We do not have to solve it with this approach I outlined above, but irrespective of the approach we resort to, we need to make while loop analysis safe, which it isn't in its current state. 

Meanwhile, I suggest reverting this PR to make while loop analysis safe.

Reply:
At this point, perhaps PRing the custom-call transform is not realistic. Neuron will have to just continue to use internal while loop unroller. 

@shraiysh you can revert the PR. Perhaps you can keep the while loop config PR portion as that is quite useful.